kotlin/internal/jvm/compile.bzl

```diff
@@ -237,81 +267,174 @@ def kt_jvm_compile_action(ctx, rule_kind, output_jar, compile_jar):
             },
-            mnemonic = "KotlinCompileAbi",
+            build_java = False,
+            build_kotlin = False,
+            mnemonic = "KotlinKapt",
         )
-        if not srcs.java:
-            compile_jar = kt_compile_jar
+        generated_src_jars.append(kapt_generated_src_jar)
+        output_jars.append(kapt_generated_class_jar)
+        kt_stubs_for_java.append(JavaInfo(compile_jar=kapt_generated_stub_jar, output_jar=kapt_generated_stub_jar, neverlink = True))
+
+    java_infos = []
+    # Build Kotlin
+    if srcs.kt or srcs.src_jars:
+        kt_runtime_jar = ctx.actions.declare_file(ctx.label.name + "-kt.jar")
+        if toolchains.kt.experimental_use_abi_jars:
+            kt_compile_jar = ctx.actions.declare_file(ctx.label.name + "-kt.abi.jar")
+            outputs = {
+                "output": kt_runtime_jar,
+                "abi_jar": kt_compile_jar,
+            }
         else:
-            java_compile_jar = ctx.actions.declare_file(ctx.label.name + "-java.abi.jar")
-            java_info = java_common.compile(
-                ctx,
-                source_files = srcs.java,
-                output = java_compile_jar,
-                deps = compile_deps.deps + [JavaInfo(compile_jar=kt_compile_jar, output_jar=kt_compile_jar)],
-                java_toolchain = toolchains.java,
-                host_javabase = toolchains.java_runtime,
-            )
-            compile_jar = ctx.actions.declare_file(ctx.label.name + ".abi.jar")
-            _fold_jars_action(
-                ctx,
-                rule_kind = rule_kind,
-                output_jar = compile_jar,
-                action_type = "Abi",
-                input_jars = [
-                    kt_compile_jar,
-                    java_common.run_ijar(
-                        ctx.actions,
-                        target_label = ctx.label,
-                        jar = java_compile_jar,
-                        java_toolchain = toolchains.java,
-                    ),
-                ],
-            )
+            kt_compile_jar = kt_runtime_jar
+            outputs = {
+                "output": kt_runtime_jar,
+            }
 
-    _run_kt_builder_action(
-        ctx = ctx,
+        _run_kt_builder_action(
+            ctx = ctx,
+            rule_kind = rule_kind,
+            toolchains = toolchains,
+            dirs = dirs,
+            srcs = srcs,
+            generated_src_jars = generated_src_jars,
+            friend = friend,
+            compile_deps = compile_deps,
+            annotation_processors = [],
+            transitive_runtime_jars = transitive_runtime_jars,
+            plugins = plugins,
+            outputs = outputs,
+            build_java = False,
+            build_kotlin = True,
+            mnemonic = "KotlinCompile",
+        )
+
+        compile_jars.append(kt_compile_jar)
+        output_jars.append(kt_runtime_jar)
+        if not annotation_processors:
```

Comment:
Having KAPT as a separate action means `kotlinc` and `JavaBuilder` can run in parallel. If there are no annotation processors then it has to happen in sequence.

What we could consider doing is to always run KAPT and when there are no annotation processors just set `aptMode=stubs`

Stub generation seems to be super fast comparatively. It would simplify the code and allow `CompileKotlin` and `JavaBuilder` to always run in parallel which is probably good for small incremental builds.


```
build //src/test/data/jvm/kapt:ap_kotlin_mixed

info: [kapt] Java stub generation took 23 ms
info: [kapt] Annotation processing took 270 ms
```

Reply:
KAPT is pretty neat in that it generates compiled stubs for the Kotlin code, which the Java code can then compile against.

Right now we have:-

If there are annotation processors
  - Run KAPT -> generates: kapt-gen-stub.jar (Kotlin Stubs), kapt-gen-java-src.jar (Annotation processor gen-src)
  - In parallel
    Run Kotlinc -> inputs: src/main/**/*.kt (to compile), src/main/**/*.java + kapt-gen-java-src.jar (for Java symbol resolution)
  Run Javac -> inputs src/main/**/*.java + kapt-gen-java-src.jar (to compile), kapt-gen-stub.jar (for Kotlin symbol resolution)

With no annotation processors
  In series run
    - kotlinc -> inputs: src/main/**/*.kt (to compile), src/main/**/*.java (Java symbol resolution) outputs -> output.jar, abi.jar
    - javac -> inputs: inputs src/main/**/*.java (to compile), abi.jar (kotlin symbol resolution)

The idea I had above was for the use case without annotation processors to run KAPT anyway, but with aptMode=stubs which seems fast. It adds an extra action but means that we don't need to wait for full Kotlinc before starting Java compilation. Javac can run in parallel with Kotlinc using the kapt-gen-stubs.jar on its classpath rather than the abi.jar which is probably faster since ABI jar has to go along with a full compile and better caching, i.e: stubs don't include what Java would consider implementation details, e.g: inline method changes, plus, if you're not running in ABI mode then the caching is worse as Java compiles against the full jar rather than the ABI jar.

Let me know if you want me to pursue this change in the PR, I'm quite happy to and can run performance tests against out repo (we have a bunch of dashboards tracking different cases in our repo, ABI changes, non-ABI changes, resource changes etc).

Reply:
Let's layer it into another pr -- this one moves a lot around. I'm optimistic -- an extra action should be less painful with multiplexing.